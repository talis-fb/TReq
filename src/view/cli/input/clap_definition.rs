use clap::{Arg, ArgAction, Command};

pub fn root_command() -> Command {
    let mut app = Command::new("treq");
    app = add_main_input_values_request(app);

    app = app
        .subcommand(
            [Command::new("GET")]
                .into_iter()
                .map(add_main_input_values_request)
                .map(add_requests_flag)
                .map(add_utils_flag)
                .next()
                .unwrap(),
        )
        .subcommand(
            [Command::new("POST")]
                .into_iter()
                .map(add_main_input_values_request)
                .map(add_requests_flag)
                .map(add_utils_flag)
                .next()
                .unwrap(),
        );

    app
}

fn add_main_input_values_request(command: Command) -> Command {
    command.arg(
        Arg::new("inputs")
            .value_name("inputs")
            .required(true)
            .num_args(1..)
            .help("All entrys"),
    )
}

fn add_requests_flag(command: Command) -> Command {
    command.arg(
        Arg::new("save-as")
            .long("save-as")
            .value_name("SAVE_NAME")
            .help("Before submit request, save it as a request permanent"),
    )
}

fn add_saved_requests_flag(command: Command) -> Command {
    command.arg(
        Arg::new("save")
            .long("save")
            .action(ArgAction::SetTrue)
            .help("Before submit request, save changes"),
    )
}

fn add_utils_flag(command: Command) -> Command {
    command.arg(
        Arg::new("no-color")
            .long("no-color")
            .action(ArgAction::SetTrue)
            .help("Print output in terminal without colors (default behavior if piped)"),
    )
}

// For save commands
fn add_requests_method_arg(command: Command) -> Command {
    command.arg(Arg::new("method"))
}
fn add_requests_url_arg(command: Command) -> Command {
    command.arg(
        Arg::new("url_manual")
            .short('u')
            .long("url")
            .value_name("URL")
            .help("Sets a url manual"),
    )
}
