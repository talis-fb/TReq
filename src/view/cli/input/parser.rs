use std::collections::HashMap;
use std::str::FromStr;

use anyhow::{Error, Result};
use clap::ArgMatches;
use serde_json::Value;

use crate::app::services::request::entities::{OptionalRequestData, METHODS};
use crate::view::cli::commands::CliCommand;
use crate::view::cli::validators;

pub fn parse_clap_input_to_commands(args: ArgMatches) -> Result<Vec<CliCommand>> {
    if args.subcommand().is_none() {
        let inputs = get_inputs_from_clap_matches(&args)?;
        let (url, extra_inputs) = inputs
            .split_first()
            .ok_or(Error::msg("No inputs"))
            .and_then(|(url, rest)| match validators::is_url(url) {
                true => Ok((url, rest)),
                false => Err(Error::msg(format!("Invalid URL: {url}"))),
            })?;

        let mut optional_request = parse_list_of_data_to_request_data(extra_inputs.to_vec())?;
        optional_request.url = Some(url.to_string());

        optional_request.method = optional_request.method.or_else(|| {
            if optional_request.body.is_some() {
                Some(METHODS::POST)
            } else {
                Some(METHODS::GET)
            }
        });

        let mut commands = Vec::new();

        let has_save_as_flag = args.get_one::<String>("save-as");
        if let Some(request_name) = has_save_as_flag {
            commands.push(CliCommand::SaveRequest {
                request_data: optional_request.clone(),
                request_name: request_name.clone(),
                check_exists_before: false,
            })
        }

        let request = optional_request.to_request_data();
        commands.push(CliCommand::SubmitRequest { request });

        return Ok(commands);
    }

    let subcommand = args.subcommand().unwrap();

    match subcommand {
        ("GET" | "POST" | "PUT" | "DELETE" | "HEAD" | "PATCH", matches) => {
            let inputs = get_inputs_from_clap_matches(matches)?;
            let method = METHODS::from_str(subcommand.0)?;

            let (url, extra_inputs) = inputs
                .split_first()
                .ok_or(Error::msg("No inputs"))
                .and_then(|(url, rest)| match validators::is_url(url) {
                    true => Ok((url, rest)),
                    false => Err(Error::msg(format!("Invalid URL: {url}"))),
                })?;

            let mut optional_request = parse_list_of_data_to_request_data(extra_inputs.to_vec())?;
            optional_request.url = Some(url.to_string());
            optional_request.method = Some(method);

            optional_request.body = optional_request.body.and_then(|body| {
                if method == METHODS::GET {
                    None
                } else {
                    Some(body)
                }
            });

            let has_raw_body_input_flag = matches.get_one::<String>("raw");
            if let Some(raw_body) = has_raw_body_input_flag {
                if optional_request.body.is_some() {
                    return Err(Error::msg(
                        "You can't use --raw and --body at the same time".to_string(),
                    ));
                }

                let _: Value =
                    serde_json::from_str(raw_body).map_err(|err| Error::msg(format!("{err}")))?;

                optional_request.body = Some(raw_body.to_string());
            }

            let mut commands = Vec::new();

            let has_save_as_flag = matches.get_one::<String>("save-as");
            if let Some(request_name) = has_save_as_flag {
                commands.push(CliCommand::SaveRequest {
                    request_data: optional_request.clone(),
                    request_name: request_name.clone(),
                    check_exists_before: false,
                })
            }

            let request = optional_request.to_request_data();

            commands.push(CliCommand::SubmitRequest { request });

            Ok(commands)
        }
        ("edit", matches) => {
            let inputs = get_inputs_from_clap_matches(matches)?;
            let (name_saved_request, extra_inputs) =
                inputs.split_first().ok_or(Error::msg("No inputs"))?;
            let mut optional_request_data =
                parse_list_of_data_to_request_data(extra_inputs.to_vec())?;

            let mut commands = Vec::new();

            let has_manual_url_flag = matches.get_one::<String>("url_manual");
            if let Some(url) = has_manual_url_flag {
                optional_request_data.url = Some(url.clone());
            }

            let has_manual_method_flag = matches.get_one::<String>("method_manual");
            if let Some(method) = has_manual_method_flag {
                optional_request_data.method = Some(METHODS::from_str(method)?);
            }

            commands.push(CliCommand::SaveRequest {
                request_data: optional_request_data.clone(),
                request_name: name_saved_request.to_string(),
                check_exists_before: true,
            });

            let has_save_as_flag = matches.get_one::<String>("save-as");
            if let Some(request_name) = has_save_as_flag {
                commands.push(CliCommand::SaveRequest {
                    request_data: optional_request_data,
                    request_name: request_name.clone(),
                    check_exists_before: false,
                })
            }

            Ok(commands)
        }
        ("rename", matches) => {
            let inputs = get_inputs_from_clap_matches(matches)?;
            let old_name = inputs[0];
            let new_name = inputs[1];
            Ok(Vec::from([CliCommand::RenameSavedRequest {
                request_name: old_name.to_string(),
                new_name: new_name.to_string(),
            }]))
        }
        ("remove", matches) => {
            let inputs = get_inputs_from_clap_matches(matches)?;
            let request_name = inputs[0].to_string();
            Ok(Vec::from([CliCommand::RemoveSavedRequest { request_name }]))
        },
        ("ls", _) => {
            Ok(Vec::from([CliCommand::ShowRequests]))
        },
        ("run", matches) => {
            let inputs = get_inputs_from_clap_matches(matches)?;

            let request_name = inputs[0].to_string();
            let extra_inputs = &inputs[1..];

            let mut commands = Vec::new();

            let mut optional_request_data =
                parse_list_of_data_to_request_data(extra_inputs.to_vec())?;

            let has_manual_url_flag = matches.get_one::<String>("url_manual");
            if let Some(url) = has_manual_url_flag {
                optional_request_data.url = Some(url.clone());
            }

            let has_manual_method_flag = matches.get_one::<String>("method_manual");
            if let Some(method) = has_manual_method_flag {
                optional_request_data.method = Some(METHODS::from_str(method)?);
            }

            let has_save_current_flag = matches.get_one::<bool>("save");
            if let Some(&true) = has_save_current_flag {
                commands.push(CliCommand::SaveRequest {
                    request_data: optional_request_data.clone(),
                    request_name: request_name.clone(),
                    check_exists_before: true,
                })
            }

            let has_save_as_flag = matches.get_one::<String>("save-as");
            if let Some(new_request_name) = has_save_as_flag {
                commands.push(CliCommand::SaveRequestWithBaseRequest {
                    request_name: new_request_name.clone(),
                    base_request_name: request_name.clone(),
                    request_data: optional_request_data.clone(),
                    check_exists_before: false,
                })
            }

            commands.push(CliCommand::SubmitSavedRequest {
                request_name,
                request_data: optional_request_data.clone(),
            });

            Ok(commands)
        }
        _ => Err(Error::msg("No valid subcommand")),
    }
}

fn get_inputs_from_clap_matches(args: &ArgMatches) -> Result<Vec<&String>> {
    Ok(args
        .get_many::<String>("inputs")
        .ok_or(Error::msg("No inputs at command"))?
        .collect())
}

fn parse_list_of_data_to_request_data<'a>(
    inputs: impl IntoIterator<Item = &'a String>,
) -> Result<OptionalRequestData> {
    let mut request = OptionalRequestData::default();
    let mut body_data_values = HashMap::new();

    inputs.into_iter().for_each(|input| {
        if validators::is_header_input(input) {
            let (key, value) = input.split_once(':').unwrap();
            request
                .headers
                .get_or_insert(HashMap::new())
                .insert(key.to_owned(), value.to_owned());
        } else if validators::is_body_data_input(input) {
            let (key, value) = input.split_once('=').unwrap();
            body_data_values.insert(key.to_owned(), value.to_owned());
        }
    });

    if !body_data_values.is_empty() {
        request.body = {
            let raw_json = serde_json::to_string(&body_data_values)
                .map_err(|err| Error::msg(format!("Error to parse data body to json: {}", err)))?;

            Some(raw_json)
        }
    }

    Ok(request)
}
