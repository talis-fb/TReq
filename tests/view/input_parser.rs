#![allow(non_snake_case)]

use insta::assert_yaml_snapshot as assert_snapshot;
// use treq::app::services::request::entities::METHODS;
// use treq::view::cli::commands::CliCommand;
use treq::view::cli::input::clap_definition::root_command;
use treq::view::cli::input::parser::parse_clap_input_to_commands;

#[test]
fn should_parse_to_normal_GET_submit_without_passing_method_as_subcommand_and_no_body() {
    let input = "treq url.com";
    let matches = root_command().get_matches_from(input.split_whitespace());
    let result = parse_clap_input_to_commands(matches).unwrap();

    assert!(result.len() == 1);
    assert_snapshot!(result);
}

#[test]
fn should_parse_to_normal_POST_submit_without_passing_method_as_subcommand_but_passing_some_body_data() {
    let input = "treq url.com Hello=World";
    let matches = root_command().get_matches_from(input.split_whitespace());
    let result = parse_clap_input_to_commands(matches).unwrap();

    assert!(result.len() == 1);
    assert_snapshot!(result);
}

#[test]
fn should_parse_all_methods_subcommands_to_normal_submits() {
    let all_methods = [
        "GET",
        "POST",
        "PUT",
        "DELETE",
        "HEAD",
        "PATCH",
    ];

    let inputs = all_methods
        .iter()
        .map(|method| format!("treq {} url.com", method))
        .collect::<Vec<_>>();

    inputs.iter().for_each(|input| {
        let matches = root_command().get_matches_from(input.split_whitespace());
        let result = parse_clap_input_to_commands(matches).unwrap();
        assert!(result.len() == 1);
        assert_snapshot!(result);
    });
}

#[test]
fn should_parse_same_way_with_or_without_protocol_in_url() {
    let input1 = "treq url.com";
    let input2 = "treq http://url.com";

    let matches1 = root_command().get_matches_from(input1.split_whitespace());
    let matches2 = root_command().get_matches_from(input2.split_whitespace());

    let result1 = parse_clap_input_to_commands(matches1).unwrap();
    let result2 = parse_clap_input_to_commands(matches2).unwrap();

    assert_eq!(result1, result2);
}

#[test]
fn should_error_if_no_input() {
    let input = "treq";
    let matches = root_command().get_matches_from(input.split_whitespace());
    let result = parse_clap_input_to_commands(matches);
    assert!(result.is_err());
}


// TODO Fix it in parser using regex validator
#[test]
fn should_error_if_subcommand_is_not_supported() {
    // let input = "treq unknown url.com";
    // let matches = root_command().get_matches_from(input.split_whitespace());
    // let result = parse_clap_input_to_commands(matches);
    // assert!(result.is_err());
}



// TODO: The body value has changed in ever iteration, once it's a HashMap and order of elements is
// changed. This test will be fixed

// #[test]
// fn should_persist_snapshot_to_edit() {
//     let input = "treq edit url.com";
//     let matches = root_command().get_matches_from(input.split_whitespace());
//     let result = parse_clap_input_to_commands(matches).unwrap();
//     assert_snapshot!(result);
//
//     let input_with_body = "treq edit url.com Hello=World name=Bob";
//     let matches = root_command().get_matches_from(input_with_body.split_whitespace());
//     let result = parse_clap_input_to_commands(matches).unwrap();
//     assert_snapshot!(result);
//
//     let input_with_body_with_header = "treq edit url.com Auth:'Bearer token' Hello=World name=Bob";
//     let matches = root_command().get_matches_from(input_with_body_with_header.split_whitespace());
//     let result = parse_clap_input_to_commands(matches).unwrap();
//     assert_snapshot!(result);
//
//     let input_with_other_url_and_method = "treq edit url.com --method GET --url http://some_another_api.com";
//     let matches = root_command().get_matches_from(input_with_other_url_and_method.split_whitespace());
//     let result = parse_clap_input_to_commands(matches).unwrap();
//     assert_snapshot!(result);
// }
