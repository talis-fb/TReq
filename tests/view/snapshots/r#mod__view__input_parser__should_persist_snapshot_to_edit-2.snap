---
source: tests/view/input_parser.rs
expression: result
---
- SaveRequest:
    request_name: url.com
    request_data:
      url: ~
      method: ~
      headers: ~
      body: "{\"name\":\"Bob\",\"Hello\":\"World\"}"
    check_exists_before: true

